hudi-metastore/src/main/java/org/apache/hudi/common/table/HoodieTableMetastoreClient.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.hudi.metastore.thrift.Table;
+import org.apache.hudi.common.table.timeline.HoodieMetastoreBasedTimeline;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import java.util.List;
+import java.util.Set;
+
+public class HoodieTableMetastoreClient extends HoodieTableMetaClient {
+  private static final Logger LOG = LogManager.getLogger(HoodieTableMetastoreClient.class);
+
+  private String databaseName;
+  private String tableName;
+  private Table table;
+  private HoodieMetastoreClient metastoreClient;
+
+  public HoodieTableMetastoreClient(Configuration conf, ConsistencyGuardConfig consistencyGuardConfig, FileSystemRetryConfig fileSystemRetryConfig,
+      String databaseName, String tableName, HoodieMetastoreConfig config) {
+    super(conf, config.getString(HoodieWriteConfig.BASE_PATH), false, consistencyGuardConfig, Option.of(TimelineLayoutVersion.CURR_LAYOUT_VERSION),
+        config.getString(HoodieTableConfig.PAYLOAD_CLASS_NAME), fileSystemRetryConfig);
+    if (databaseName == null || tableName == null) {
+      throw new HoodieException("The database and table name have to be specified");
+    }
+    this.databaseName = databaseName;
+    this.tableName = tableName;
+    this.metastoreConfig = config;
+    this.metastoreClient = RetryingHoodieMetastoreClient.getProxy(config);
+    this.table = initOrGetTable(databaseName, tableName, config);
+    // TODO: transfer table parameters to table config
+    this.tableConfig = new HoodieTableConfig();
+    tableConfig.setTableVersion(HoodieTableVersion.THREE);
+    tableConfig.setAll(config.getProps());
+  }
+
+  private Table initOrGetTable(String db, String tb, HoodieMetastoreConfig config) {
+    Table table;
+    try {
+      table = metastoreClient.getTable(databaseName, tableName);
+    } catch (HoodieException e) {
+      if (e.getCause() instanceof NoSuchObjectException) {
+        LOG.info(String.format("Table %s.%s doesn't exist, will create it.", db, tb));
+        table = new Table();
+        table.setDbName(db);
+        table.setTableName(tb);
+        table.setLocation(config.getString(HoodieWriteConfig.BASE_PATH));
+        table.setOwner("");
+        table.setTableType(config.getString(HoodieTableConfig.TYPE.key()));
+        metastoreClient.createTable(table);
+        table = metastoreClient.getTable(databaseName, tableName);
+      } else {
+        throw e;
+      }
+    }
+    return table;
+  }
+
+  /**
+   * @return Hoodie Table Type
+   */
+  public HoodieTableType getTableType() {
+    return HoodieTableType.valueOf(table.getTableType());
+  }
+
+  /**
+   * Get the active instants as a timeline.
+   *
+   * @return Active instants timeline
+   */
+  public synchronized HoodieActiveTimeline getActiveTimeline() {
+    if (activeTimeline == null) {
+      activeTimeline = new HoodieMetastoreBasedTimeline(this, metastoreConfig);
+    }
+    return activeTimeline;
+  }
+
+  /**
+   * Reload ActiveTimeline and cache.
+   *
```

Comment:
oh, sorry.   I didn't see any cache code in this function.

Reply:
will delete